### ISSUE TITLE
Doesn't properly handle write-masked SSBO stores when converting to SPIR-V

### ISSUE THREAD

**Issue by jekstrand** [KhronosGroup/glslang#94]
Consider the following shader:

```
layout(location = 0) in vec2 data;
layout(set = 0, binding = 0, std140) buffer Storage {
   vec4 arr[];
} ssbo;

void main()
{
   if (gl_VertexID % 2 == 0) {
      ssbo.arr[gl_VertexID / 2].xz = data;
   } else {
      ssbo.arr[gl_VertexID / 2].yw = data;
   }
}
```

The two SSBO stores should only alter the components of the SSBO specified by the swizzle.  I can invoke this shader as a SIMD8 shader on a Broadwell in such a way that one SIMD8 shader invocation will take vertices with an even gl_VertexID and another will take the odd ones.  This means that one EU (execution unit) will only execute the first case and the other will only execute the second case.  GLSLang converts this shader to the following SPIR-V:

```
                              Source GLSL 430
                              SourceExtension  "GL_GOOGLE_cpp_style_line_directive"
                              SourceExtension  "GL_GOOGLE_include_directive"
                              Capability Shader
               1:             ExtInstImport  "GLSL.std.450"
                              MemoryModel Logical GLSL450
                              EntryPoint Vertex 4  "main"
                              Name 4  "main"
                              Name 8  "gl_VertexID"
                              Name 20  "Storage"
                              MemberName 20(Storage) 0  "arr"
                              Name 22  "ssbo"
                              Name 27  "data"
                              Name 40  "gl_InstanceID"
                              Decorate 8(gl_VertexID) BuiltIn VertexId
                              Decorate 19 ArrayStride 16
                              MemberDecorate 20(Storage) 0 Offset 0
                              Decorate 20(Storage) BufferBlock
                              Decorate 22(ssbo) DescriptorSet 0
                              Decorate 22(ssbo) Binding 0
                              Decorate 27(data) Location 0
                              Decorate 40(gl_InstanceID) BuiltIn InstanceId
               2:             TypeVoid
               3:             TypeFunction 2
               6:             TypeInt 32 1
               7:             TypePointer Input 6(int)
  8(gl_VertexID):      7(ptr) Variable Input
              10:      6(int) Constant 2
              12:      6(int) Constant 0
              13:             TypeBool
              17:             TypeFloat 32
              18:             TypeVector 17(float) 4
              19:             TypeRuntimeArray 18(fvec4)
     20(Storage):             TypeStruct 19
              21:             TypePointer Uniform 20(Storage)
        22(ssbo):     21(ptr) Variable Uniform
              25:             TypeVector 17(float) 2
              26:             TypePointer Input 25(fvec2)
        27(data):     26(ptr) Variable Input
              29:             TypePointer Uniform 18(fvec4)
40(gl_InstanceID):      7(ptr) Variable Input
         4(main):           2 Function None 3
               5:             Label
               9:      6(int) Load 8(gl_VertexID)
              11:      6(int) SMod 9 10
              14:    13(bool) IEqual 11 12
                              SelectionMerge 16 None
                              BranchConditional 14 15 33
              15:               Label
              23:      6(int)   Load 8(gl_VertexID)
              24:      6(int)   SDiv 23 10
              28:   25(fvec2)   Load 27(data)
              30:     29(ptr)   AccessChain 22(ssbo) 12 24
              31:   18(fvec4)   Load 30
              32:   18(fvec4)   VectorShuffle 31 28 4 1 5 3
                                Store 30 32
                                Branch 16
              33:               Label
              34:      6(int)   Load 8(gl_VertexID)
              35:      6(int)   SDiv 34 10
              36:   25(fvec2)   Load 27(data)
              37:     29(ptr)   AccessChain 22(ssbo) 12 35
              38:   18(fvec4)   Load 37
              39:   18(fvec4)   VectorShuffle 38 36 0 4 2 5
                                Store 37 39
                                Branch 16
              16:             Label
                              Return
                              FunctionEnd
```

As you can see, in both cases, it does a load, a vector-shuffle, and a store.  This creates a race condition where both EU's could, in theory, do the load at the same time, execute the vector-shuffle, and then both store all 4 components.  Whoever gets to the store first wins the race and the other thread's data gets stompped.

We found a similar bug in the GLSL compiler in upstream mesa recently which was causing it to fail compute shader conformance tests.

**Comment by Themaister** [comment]
I assume the correct way to do this would be access chaining the SSBO vector down to component level and do two independent writes?

**Comment by jekstrand** [comment]
That would be certainly be a correct implementation.

**Comment by johnkslang** [comment]
Yes, the intention is these should be independent writes to the scalar components involved, with no writes to the other components.  There is code already written, on its way, that creates the right infrastructure and does it correctly for a single component.

**Comment by cmarcelo** [comment]
> There is code already written, on its way, that creates the right infrastructure and does it correctly for a single component.

@johnkslang has the mentioned infrastructure to fix this landed?

**Comment by greg-lunarg** [comment]
Closing in deference to #2518.

**Comment by greg-lunarg** [comment]
Fixed by #2702
